odoo/addons/base/models/ir_binary.py

```diff
@@ -0,0 +1,247 @@
+from odoo.http import Stream, request
+from odoo.tools import file_open, replace_exceptions
+from odoo.tools.image import image_process, image_guess_size_from_field_name
+from odoo.tools.mimetypes import guess_mimetype, get_extension
+
+
+DEFAULT_PLACEHOLDER_PATH = 'web/static/img/placeholder.png'
+_logger = logging.getLogger(__name__)
+
+
+class IrBinary(models.AbstractModel):
+    _name = 'ir.binary'
+    _description = "File streaming helper model for controllers"
+
+    def _find_record(
+            self, xmlid=None, res_model='ir.attachment', res_id=None,
+            access_token=None,
+    ):
+        """
+        Find and return a record either using an xmlid either a model+id
+        pair. This method is an helper for the ``/web/content`` and
+        ``/web/image`` controllers and should not be used in other
+        contextes.
+
+        :param Optional[str] xmlid: xmlid of the record
+        :param Optional[str] res_model: model of the record,
+            ir.attachment by default.
+        :param Optional[id] res_id: id of the record
+        :param Optional[str] access_token: access token to use instead
+            of the access rights and access rules.
+        :returns: single record
+        :raises MissingError: when no record was found.
+        """
+        record = None
+        if xmlid:
+            record = self.env.ref(xmlid, False)
+        elif res_id is not None and res_model in self.env:
+            record = self.env[res_model].browse(res_id).exists()
+        if not record:
+            raise MissingError(f"No record found for xmlid={xmlid}, res_model={res_model}, id={res_id}")
+
+        if record._name == 'ir.attachment':
+            record = record.validate_access(access_token)
+
+        return record
+
+    def _record_to_stream(self, record, field_name):
+        """
+        Low level method responsible for the actual conversion from a
+        model record to a stream. This method is an extensible hook for
+        other modules. It is not meant to be directly called from
+        outside or the ir.binary model.
+
+        :param record: the record where to load the data from.
+        :param str field_name: the binary field where to load the data
+            from.
+        :rtype: odoo.http.Stream
+        """
+        if record._name == 'ir.attachment' and field_name in ('raw', 'datas', 'db_datas'):
+            return Stream.from_attachment(record)
+
+        field_def = record._fields[field_name]
+
+        # fields.Binary(attachment=False) or compute/related
+        if not field_def.attachment or field_def.compute or field_def.related:
+            return Stream.from_binary_field(record, field_name)
+
+        # fields.Binary(attachment=True)
+        field_attachment = self.env['ir.attachment'].sudo().search(
+            domain=[('res_model', '=', record._name),
+                    ('res_id', '=', record.id),
+                    ('res_field', '=', field_name)],
+            limit=1)
```

Comment:
@tivisse this is basically a copy from 62c9dedafda0cbdc618641eaa80c49535ab0b5f6 do you think using `record.env`